gateway-service/src/main/resources/application.yml

```diff
@@ -39,4 +39,9 @@ spring:
     properties:
       hibernate:
         format_sql: true
-        use_sql_comments: true
\ No newline at end of file
+        use_sql_comments: true
+
+  mvc:
```

Comment:
그러면 코드가 설정에 의존성이 생긴다는 것인가요? 404면 설정과 관계없이 에러가 발생되야 하는것 아닌가요?

Reply:
아뇨 코드가 설정에 의존성이 생긴다는 것이 무슨 말씀일까요....? 설정과 관계없이 DispatcherServlet에선 요청 url 에 매핑되는 핸들러가 없다면 404에러는 기본적으로 발생합니다....(404에러 발생과 NoHandlerFoundException 발생은 별개로 생각해야 합니다)
제가 한 건 스프링부트 Web MVC 기본설정을 변경시킨 것이죠.,......
1. spring.mvc.throw-exception-if-no-handler-found: true
아래는 적절한 핸들러가 존재하지 않을 경우  DispatcherServlet에서 호출되는 함수입니다. 

<DispatcherServlet의 주요부분>
`
* Set whether to throw a NoHandlerFoundException when no Handler was found for this request.
	 * This exception can then be caught with a HandlerExceptionResolver or an
	 * {@code @ExceptionHandler} controller method.
	 * <p>Note that if {@link org.springframework.web.servlet.resource.DefaultServletHttpRequestHandler}
	 * is used, then requests will always be forwarded to the default servlet and a
	 * NoHandlerFoundException would never be thrown in that case.
	 * <p>Default is "false", meaning the DispatcherServlet sends a NOT_FOUND error through the
	 * Servlet response.
	 * @since 4.0
	 */
	public void setThrowExceptionIfNoHandlerFound(boolean throwExceptionIfNoHandlerFound) {
		this.throwExceptionIfNoHandlerFound = throwExceptionIfNoHandlerFound;
	}

/**
	 * No handler found -> set appropriate HTTP response status.
	 * @param request current HTTP request
	 * @param response current HTTP response
	 * @throws Exception if preparing the response failed
	 */
	protected void noHandlerFound(HttpServletRequest request, HttpServletResponse response) throws Exception {
		if (pageNotFoundLogger.isWarnEnabled()) {
			pageNotFoundLogger.warn("No mapping for " + request.getMethod() + " " + getRequestUri(request));
		}
		if (this.throwExceptionIfNoHandlerFound) {
			throw new NoHandlerFoundException(request.getMethod(), getRequestUri(request),
					new ServletServerHttpRequest(request).getHeaders());
		}
		else {
			response.sendError(HttpServletResponse.SC_NOT_FOUND);
		}
	}`

위 메서드에서도 나타나지만 throwExceptionIfNoHandlerFound는 디폴트값이 false 이기 때문에 noHandlerFound 메서드에서 두번째 조건문이 아닌 아래 else 문을 타게 되는데요, 제가 원하는 것은 DispatcherServlet에서 바로 404에러를 응답으로 반환시키는 것이 아니라, NoHandlerFoundException을 발생시켜 제가 작성한 @ExceptionHandler(NoHandlerFoundException) 메서드에서 포착되게 하여 커스텀한 에러처리 로직을 타도록 하는 것이죠. 이를 위해서 프로퍼티파일을 통해 throwExceptionIfNoHandlerFound를 true로 변경하는 설정을 추가한 것입니다. 

2. spring.resources.add-mappings: false
스프링부트의 기본 Web MVC 설정에 의해 우선순위가 가장 낮게 등록되어있는 매핑 빈은 SimpleUrlHandlerMapping으로, DispatcherServlet에게 ResourceHttpRequestHandler (정적자원처리담당)를 리턴합니다, 여기에서도 처리할 수 없는 요청이 되면 또 404에러를 반환하죠. 역시나 구현하고자 하는건 커스텀 에러 처리 핸들링 로직에 의해 예외가 처리되도록 하는 것이기 때문에, url 매핑이 안된 경우 정적자원처리핸들러까지 들어가지 않도록 해당 설정을 추가하였습니다.

Reply:
네 ㅎㅎ 제가 언급한 관점은 왜 여러 ExceptionHandler 중에 왜 저 녀석만 추가 설정이 있는지에 대해서 입니다. 보면 @Override 해서 만든 여러개의 메소드들 중에 왜 그 녀석만 설정에 영향을 받는지에 대한 부분이 "**일관성**"이 흐트러지는 부분이 생기는 건가? 에 대한 물음이었거든요 
저 기능을 사용하기 위해 필요한 건데 이걸 왜 생각해야 하지? 라고 질문하게 된다면 즉 기술적인 부분이나 해당 Custom Exception을 발생하는데 필요한 부분이 라는 것은 매우 설명을 잘 해주어서 해당 부분을 추가해야하는 당위성은 이해 되었습니다. 다만 일관적으로 common 한 성격이라기 보다는 특정 @Override 한 메소드에 특이점은 아닌지 그리고 그 부분을 다른 개발자게에 넘겨야 할때 다른 사람에게 "인수인계 or 설명" 해야 하는 추가 비용을 묻고 싶지 않아서 였어요 즉 코드만 보고 다 알 수 있으면 설정파일을 확인 해야 하는 수고스러움을 덜 할 수 있다는 것이겠죠 
지금은 이제 처음 생긴 Exception 쪽 설정이지만 이 코드가 생물과 같이 시간이 흐르면서 변화할 텐데 그 이후에도 계속 약간의 예외성을 띄고 있는 설정들이 계속 나온다면? 어쨌든 개발 문서나 API 설명을 꼼꼼히 읽어서 모든 사람이 착해서 그런 수고스러움이 적으면 좋겠지만 일단 그 부분을 잘 모르는 상황에서는 친절함이 부족한 부분이 있다는 거죠 
개발자도 여러 부류가 있지만 지금과 같은 Spring Framework 에 있는 부분을 개발한 Core 개발자가 있고 그것을 이용만 하는 비즈니스 개발자로 분류할 때 비즈니스 개발자가 쓰기에 불편하게 만들지 않았니?? 라는 질문을 Core 개발자에게 하기 위함이에요 Spring framework 가 만능도 아니고 잘만들었다고 100% 신뢰할 수는 없기 때문에 왜 이녀석은 다른 메소드와 달리 설정에 추가 값을 넣어야 하니? 라는 물음을 한다는 거죠 차라리 메소드 내부에 " boolean flagCustom = false " 이런식의 기본값이 되었다면 조금 더 비즈니스개발자가 쓰기에 편하지 않냐는 것인거죠 
즉 이 부분까지 나간건 저 기능을 쓰기가 편리하게 만들어 졌니? 라는 질문이 생긴거였어요 Core 개발자 들이 물론 수준이 뛰어난 개발자들이 저 코드를 넣었겠지만 항상 옳은가? 에서는 아닐 수 있고 또 core 개발자라면 라이브러리를 제공하는 관점에서는 편리하게 제공하니? 라는 질문을 할 수 있었으면 생각하게 되었네요 
어쨌든 수정님이 부연 설명을 잘해주셔서 Custom 기능을 쓰려면 해당 설정은 추가할 수 밖에 없는거 같아서 이 PR은 마무리하고 Merge하도록 하겠습니다~~~ API나 라이브러리의 친절함이나 다른 매소드들과 일관성을 생각한 관점에서 질문이었습니다. 기능상에 질문은 아니었는데 답변이 디테일해서 이해하기 좋았어요!!!

Reply:
과거의 Spring Framework 의 경우 저는 너무 어려운게 xml 설정이었습니다. 기능이나 기술이 아닌 사람이 쓰기 좋은가? 이런 질문이었다고 하면 이해가 좀 될까요? ....